saleor/graphql/discount/mutations/voucher/voucher_update.py

```diff
@@ -33,6 +39,96 @@ class Meta:
         ]
 
     @classmethod
-    def post_save_action(cls, info: ResolveInfo, instance, cleaned_input):
+    def clean_input(cls, info: ResolveInfo, instance, data, **kwargs):
+        cls.clean_codes(data)
+        cls.clean_voucher_usage_setting(instance, data)
+        cleaned_input = super().clean_input(info, instance, data, **kwargs)
+
+        return cleaned_input
+
+    @classmethod
+    def clean_codes(cls, data):
+        if "code" in data:
+            cls._clean_old_code(data)
+
+        if "add_codes" in data:
+            cls._clean_new_codes(data)
+
+    @classmethod
+    def clean_voucher_usage_setting(cls, instance, data):
+        """Ensure single use setting is not changed if voucher was already used."""
+        if "single_use" in data and instance.single_use != data["single_use"]:
+            voucher_codes = instance.codes.all()
+            used_codes = voucher_codes.filter(
+                Exists(order_models.Order.objects.filter(voucher_code=OuterRef("code")))
+                | Exists(
+                    order_models.OrderLine.objects.filter(voucher_code=OuterRef("code"))
+                )
+                | Exists(
+                    checkout_models.Checkout.objects.filter(
+                        voucher_code=OuterRef("code")
+                    )
+                )
+            )
+            if used_codes.exists():
+                raise ValidationError(
+                    {
+                        "single_use": ValidationError(
+                            "Cannot change single use setting when any voucher code has "
+                            "already been used.",
+                            code=DiscountErrorCode.VOUCHER_ALREADY_USED.value,
+                        )
+                    }
+                )
+
+    @classmethod
+    def construct_codes_instances(
+        cls, code, codes_data, cleaned_input, voucher_instance
+    ):
+        if codes_data:
```

Comment:
Here you have different order of picking up the codes than in `VoucherCreate`. It means that it will behave differently for Create&Update.

Reply:
API doesn't allow to send `code` and `add_codes` together so the behavior will be the same

Reply:
Yeah, miss the validation of arguments!